ckanext/canada/plugins.py

```diff
@@ -440,3 +445,75 @@ def after_show(self, context, data_dict):
 
     def update_facet_titles(self, facet_titles):
         return facet_titles
+
+
+@chained_action
+def datastore_upsert(up_func, context, data_dict):
+    lc = ckanapi.LocalCKAN(username=c.user)
+    res_data = lc.action.datastore_search(
+        resource_id=data_dict['resource_id'],
+        filters={},
+        limit=1,
+    )
+    count = res_data.get('total', 0)
+    result = up_func(context, data_dict)
+
+    res_data = lc.action.datastore_search(
+        resource_id=data_dict['resource_id'],
+        filters={},
+        limit=1,
+    )
+    count = res_data.get('total', 0) - count
+
+    act.datastore_activity_create(context, {'count':count,
+                                            'activity_type': 'changed datastore',
+                                            'resource_id': data_dict['resource_id']}
+                                  )
+    return result
+
+
+@chained_action
+def datastore_delete(up_func, context, data_dict):
+    lc = ckanapi.LocalCKAN(username=c.user)
+    res = lc.action.datastore_search(
+        resource_id=data_dict['resource_id'],
+        filters=data_dict['filters'],
+        limit=1,
+    )
+    result = up_func(context, data_dict)
+    act.datastore_activity_create(context,
+                                  {'count':res.get('total', 0),
+                                   'activity_type': 'deleted datastore',
+                                   'resource_id': data_dict['resource_id']}
+                                  )
+    return result
+
+
+class CanadaActivity(p.SingletonPlugin):
+    p.implements(p.IActions)
+    p.implements(IActivity)
+
+    def get_actions(self):
+        return ({'datastore_upsert':datastore_upsert,
+                'datastore_delete': datastore_delete})
+
+    def string_icons(self, string_icons):
+        string_icons.update({'deleted datastore': 'file',
+                             'changed datastore': 'file'})
+
+    def snippet_functions(self, snippet_functions):
+        snippet_functions.update({'datastore': act.get_snippet_datastore,})
+        snippet_functions.update({'datastore_detail':
+                                  act.get_snippet_datastore_detail,})
```

Comment:
looks a little odd. Do you mean to update both with one call? If not
```python
snippet_functions['datastore_detail'] = act.get_snippet_datastore_detail
```
is more idiomatic